cpp/src/cider/benchmarks/HashMapBenchmark.cpp

```diff
@@ -0,0 +1,159 @@
+    auto curr = ++it;
+    if (curr)
+      curr->getMapped();
+  }
+
+  return map.getCollisions();
+}
+
+template <typename T>
+static std::tuple<std::vector<T>, std::vector<bool>> generateAndFillVector(
+    const size_t row_num,
+    const GeneratePattern pattern,
+    const int32_t
+        null_chance,  // Null chance for each column, -1 represents for unnullable
+                      // column, 0 represents for nullable column but all data is not
+                      // null, 1 represents for all rows are null, values >= 2 means
+                      // each row has 1/x chance to be null.
+    const T value_min = N_MIN,
+    const T value_max = N_MAX) {
+  std::vector<T> col_data(row_num);
+  std::vector<bool> null_data(row_num);
+  std::mt19937 rng(std::random_device{}());  // NOLINT
+  switch (pattern) {
+    case GeneratePattern::Sequence:
+      for (auto i = 0; i < row_num; ++i) {
+        null_data[i] = Random::oneIn(null_chance, rng) ? (col_data[i] = N_MIN, true)
+                                                       : (col_data[i] = i, false);
+      }
+      break;
+    case GeneratePattern::Random:
+      if (std::is_integral<T>::value) {
+        for (auto i = 0; i < col_data.size(); ++i) {
+          null_data[i] = Random::oneIn(null_chance, rng)
+                             ? (col_data[i] = N_MIN, true)
+                             : (col_data[i] = static_cast<T>(
+                                    Random::randInt64(value_min, value_max, rng)),
+                                false);
+        }
+      } else if (std::is_floating_point<T>::value) {
+        for (auto i = 0; i < col_data.size(); ++i) {
+          null_data[i] = Random::oneIn(null_chance, rng)
+                             ? (col_data[i] = N_MIN, true)
+                             : (col_data[i] = static_cast<T>(
+                                    Random::randFloat(value_min, value_max, rng)),
+                                false);
+        }
+      } else {
+        std::string str = "Unexpected type:";
+        str.append(typeid(T).name()).append(", could not generate data.");
+        CIDER_THROW(CiderCompileException, str);
+      }
+      break;
+  }
+  return std::make_tuple(col_data, null_data);
+}
+
+template <typename KeyType, typename HashTableType>
+static void BM_Lookup(benchmark::State& state) {
+  size_t row_num = 100;
+  KeyType value_min = std::numeric_limits<KeyType>::min();
+  KeyType value_max = std::numeric_limits<KeyType>::max();
+  size_t null_chance = 0;
+  GeneratePattern pattern = GeneratePattern::Random;
+  std::vector<KeyType> data(row_num);
+  std::vector<bool> data_null;
+  {
+    std::tie(data, data_null) =
+        value_min > value_max
+            ? generateAndFillVector<KeyType>(row_num, pattern, null_chance)
+            : generateAndFillVector<KeyType>(
+                  row_num, pattern, null_chance, value_min, value_max);
+  }
+  for (auto _ : state) {
+    for (int64_t i = state.range(0); i--;) {

```

Comment:
avoid such code. it will be unexpected if `i` is negative.

Reply:
I followed the example here.. https://github.com/google/benchmark/blob/main/docs/user_guide.md#templated-benchmarks